src/types.jl

```diff
@@ -0,0 +1,193 @@
+## pros: wrapping adds overhead, so if possible best to use Basic
+## cons: have to write methods meth(x::Basic, ...) = meth(Sym(x),...)
+
+
+## Wrapper type
+## this allows SymEngine.jl to keep track of the class of the C++ object
+
+
+## XXX Temporarary, to be replaced by get_class_from_id
+const SYMENGINE_ENUM = Dict{Int, Symbol}(0 => :Integer,
+                                         1 => :Rational,
+                                         2 => :Complex,
+                                         3 => :ComplexDouble,
+                                         11 => :Symbol,
+                                         12 => :EmptySet,
+                                         13 => :Interval,
+                                         14 => :Mul,
+                                         15 => :Add,
+                                         16 => :Pow,
+                                         19 => :Constant,
+                                         20 => :Sin,
+                                         21 => :Cos
+                                         )
+
+## Parameterized type allowing or dispatch on Julia side by type of objecton SymEngine side
+## Use as BasicType{Val{:Integer}}(...)
+## To take advantage of this, define
+## meth(x::Basic) = meth(_Sym(x))
+## and then
+## meth(x::BasicType{Val{:Integer}}) = ... or
+## meth(x::BasicNumber) = ...
+type BasicType{T} <: Number
+    x::Basic
+end
+
+SymbolicType = Union{Basic, BasicType}
+
+Basic(x::BasicType) = x.x
+
+function get_type(s::Basic)
+    ccall((:basic_get_type, :libsymengine), UInt, (Ptr{Basic},), &s)
+end
+
+function get_class_from_id(id)
+    id = string(id)
+    ccall((:basic_get_class_from_id, :libsymengine), AbstractString, (Ptr{AbstractString},), &id)
+end
+
+## Convert a Basic value into one of the BasicType values
+## XXX this needs hooking up with get_class_from_id XXX
+function Base.convert(::Type{BasicType}, val::Basic)
+    id = get_type(val)
+    # nm = get_class_from_id(id)
+    nm = haskey(SYMENGINE_ENUM, id) ? SYMENGINE_ENUM[id] : :Value  # work around until get_class_id is exposed
+    BasicType{Val{nm}}(val)
+end
+Base.convert{T}(::Type{BasicType{T}}, val::Basic) = convert(BasicType, val)
+
+## some type unions used for dispatch
+number_types = [:Integer, :Rational, :Complex]
+BasicNumber = Union{[SymEngine.BasicType{Val{i}} for i in number_types]...}
+
+op_types = [:Mul, :Add, :Pow, :Symbol, :Const]
+BasicOp = Union{[SymEngine.BasicType{Val{i}} for i in op_types]...}
+
+trig_types = [:Sin, :Cos, :Tan, :Csc, :Sec, :Cot, :ASin, :ACos, :ATan, :ACsc, :ASec, :ACot]
+BasicTrigFunction =  Union{[SymEngine.BasicType{Val{i}} for i in trig_types]...}
+
+
+
+
+
+Base.promote_rule{S<:Number}(::Type{Basic}, ::Type{S} ) = Basic
+Base.promote_rule{T<:BasicType, S<:Number}(::Type{T}, ::Type{S} ) = T
+
+Base.promote_type{T}(::Type{BasicType{Val{T}}}, ::Type{Basic}) = BasicType{Val{T}}
+Base.promote_type{T}(::Type{Basic}, ::Type{BasicType{Val{T}}}) = BasicType{Val{T}}
```

Comment:
If I add that and remove the `promote_type` lines, I get the stackoverflow
in the tests when trying to add a BasicType object to a Basic object.
However, if I also add in what should be the same:

```
Base.promote_rule{T<:BasicType}( ::Type{Basic}, ::Type{T} ) = T
```

Then the tests pass. There was a conversation (
https://groups.google.com/forum/?fromgroups#!topic/julia-users/_8T6mWJDhKk)
where the solution was to define promote_rule in a symmetric way. I don't
know why it is needed here, but must have something to do with the use of
parametric types.

On Sun, Mar 27, 2016 at 1:47 PM, Isuru Fernando <notifications@github.com>
wrote:

> In src/types.jl
> <https://github.com/symengine/SymEngine.jl/pull/19#discussion_r57532294>:
>
> > +
> > +op_types = [:Mul, :Add, :Pow, :Symbol, :Const]
> > +BasicOp = Union{[SymEngine.BasicType{Val{i}} for i in op_types]...}
> > +
> > +trig_types = [:Sin, :Cos, :Tan, :Csc, :Sec, :Cot, :ASin, :ACos, :ATan, :ACsc, :ASec, :ACot]
> > +BasicTrigFunction =  Union{[SymEngine.BasicType{Val{i}} for i in trig_types]...}
> > +
> > +
> > +
> > +
> > +
> > +Base.promote_rule{S<:Number}(::Type{Basic}, ::Type{S} ) = Basic
> > +Base.promote_rule{T<:BasicType, S<:Number}(::Type{T}, ::Type{S} ) = T
> > +
> > +Base.promote_type{T}(::Type{BasicType{Val{T}}}, ::Type{Basic}) = BasicType{Val{T}}
> > +Base.promote_type{T}(::Type{Basic}, ::Type{BasicType{Val{T}}}) = BasicType{Val{T}}
>
> I think the following missing rule is needed.
> Base.promote_rule{T<:BasicType}(::Type{T}, ::Type{Basic} ) = T
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/symengine/SymEngine.jl/pull/19/files/5f0f174ab4816a311b671e0f669b670a8b3fa0fb#r57532294>
>



-- 
John Verzani
Chair, Department of Mathematics
College of Staten Island, CUNY
verzani@math.csi.cuny.edu